### ISSUE TITLE
Track conditional narrowing between multiple variables

### ISSUE THREAD

**Issue by nadav7679** [astral-sh/ty#2496]
### Question

Hi ty team, loving your work.

I encountered quite a simple piece of logic that I can't seem to restructure in a way that passes ty check. I was wondering what is the best practice to handle it.  

The use case is something like this (also view it in this [playground](https://play.ty.dev/a94c53c4-a77d-49c7-8743-2b5e33669991)):

```python3
from typing import reveal_type

class A:
    pass

class B:
    pass

def f(
        a: A | None,
        b: B | None,
) -> A | B | tuple[A, B]:

    if a is None and b is None:
        raise

    if a is None:
        reveal_type(b) # B | None
        return b
    
    if b is None:
        return a

    return (a, b)
```

Basically, the function `f` is allowed to be triggered with either `a`, `b`, or both, but is not allowed to be triggered with two `None`s. `ty` can't recognise this logic so it treats `b` as type `B | None` and displays this error when trying to return B:

```
Return type does not match returned value: expected `A | B | tuple[A, B]`, found `B | None` (invalid-return-type) [Ln 18, Col 16]
```

I tried playing around with it and couldn't find a good solution. I was wondering what's the best practice in these kinds of cases.

Thanks!

**Comment by AlexWaygood** [comment]
It's maybe a bit weird, but we do support narrowing unions of tuples via subscript checks, and we also support narrowing of individual subscript elements, so you _could_ do something like this:

```py
from typing import reveal_type

class A:
    value: int

class B:
    value: int

def f(ab: tuple[A | None, B] | tuple[A, B | None]) -> A | B | tuple[A, B]:

    if ab[0] is None and ab[1] is None:
        raise

    if ab[0] is None:
        # `ab` is narrowed to `tuple[A | None, B]` here
        reveal_type(ab[1])  # revealed: B
        return ab[1]
    
    if ab[1] is None:
        # `ab` is narrowed to `tuple[A, B | None]` here
        return ab[0]  # revealed: A

    # at this point, `a[0]` has been narrowed to `A`
    # and `a[1]` has been narrowed to `B`,
    # (although we don't *currently* narrow the tuple as a whole
    # to `tuple[A, B]` at this location)
    return ab[0], ab[1]
```

**Comment by AlexWaygood** [comment]
(also, thanks for the kind words!)

**Comment by AlexWaygood** [comment]
The more usual thing to do is probably not to wrap your variables in a tuple, but just use an `assert` to help the type checker out:

```py
from typing import reveal_type

class A:
    pass

class B:
    pass

def f(
    a: A | None,
    b: B | None,
) -> A | B | tuple[A, B]:

    if a is None and b is None:
        raise

    if a is None:
        assert b is not None
        reveal_type(b) # B
        return b
    
    if b is None:
        return a

    return (a, b)
```

**Comment by carljm** [comment]
The key thing that ty (and all other Python type checkers I know of) don't track here is "conditional narrowing". To handle this example, we'd need to track that after `if a is None and b is None: return`, `a is not None OR b is not None`, so that later we know that `a is None` implies `b is not None`.

Some amount of this is doable, but since no other type checker does it either, it probably won't be a priority in the short term. I'll keep this issue open and track it as a "wish".

In addition to the suggestions above, you can also restructure the conditionals:

```py
if a is None:
    if b is None:
        raise
    return b
if b is None:
    return a
return (a, b)
```

**Comment by nadav7679** [comment]
Thanks both!

> we'd need to track that after if a is None and b is None: return, a is not None OR b is not None, so that later we know that a is None implies b is not None.

Yeah that's exactly what I thought. Would be nice to have (and impressive) but definitely not a must.

**Comment by Jackenmen** [comment]
Another example of where this would be useful is when a function returns a union of tuples, i.e.:
```py
def x() -> tuple[None, None] | tuple[str, str]:
    return "", ""


def y() -> str:
    a, b = x()
    if a is not None:
        return b
    return ""


def z() -> str:
    tup = x()
    if tup[0] is not None:
        return tup[1]
    return ""
```
`y()` will result in an error on `return b`, but `z()` won't since it's in a single variable:
```
Return type does not match returned value: expected `str`, found `None | str` (invalid-return-type) [Ln 8, Col 16]
```
